### ISSUE TITLE
[rllib] Policy losses that depend on trajectories of experience

### ISSUE THREAD

**Issue by nrhine1** [ray-project/ray#10189]
I want to implement a Policy loss function that requires a somewhat long trajectory of observations and actions from the same episode (e.g. a length-10 trajectory of observations and actions). 

My understanding is that the standard aggregators of experience are the `SampleBatch` and `MultiAgentBatch` objects, which don't seem to explicitly provide access to arbitrary-length trajectories of samples, at least not in any examples. However, the `SampleBatch.split_by_episodes`, `SampleBatch.rows` and `MultiAgentBatch.timeslices` look relevant.

**Comment by ericl** [comment]
The sample batches do include the episode / unroll ids (`eps_id`), so it is possible to extract contiguous unroll segments from the batch. If you further use "batch_mode: complete_episodes", then you are guaranteed to see entire episodes in the batch, and can use .split_by_episode()  to extract those.

Note that for DQN-like algorithms, you'll also want to set "replay_sequence_length: N", otherwise the training batch will be by default have random samples without any contiguous sequences.

Hope this helps!

**Comment by nrhine1** [comment]
Thanks @ericl ! Some quick details for context:

-  I'm currently using torch PPO with the default `execution_plan` 
- My goal is for the data that arrives inside the `loss_fn` to contain a set of windows (contiguous sequences) from different episodes, in order to construct, inside the loss function, an observation `Tensor` of shape `(N_windows, window_size, obs_size)`.

I interpreted your comment to mean that I could set that parameter and use `split_by_episodes` inside the `loss_fn` (or right before it, inside `TorchPolicy.learn_on_batch`, because what arrives inside the `loss_fn` is a `UsageTrackingDict`).

Even if I set `batch_mode: "complete_episodes"`, which seems to mainly affect the sampler, I don't think that guarantees that the minibatch that arrives inside the loss function will contain a set of continguous windows. I think this is because the way that `do_minibatch_sgd` creates the minibatches. As far as I can tell, the `minibatches` function will create a generator that slices out the `SampleBatch`, so the slices no longer contain what I want. This would suggest that I need to implement a different iterator instead of the `TrainOneStep` iterator..? 

I found that if I disable the shuffling inside of the `minibatches` function, and with the default minibatch sizes of 128, I ended up with two contiguous trajectories in the `postprocessed_batch` after using `split_by_episode`, one of shape `40` and another of shape `88`. All trajectories from my `env` are length-100, so I'm not totally sure what's going on.

So my current belief is that I'll need to implement a different iterator than `TrainOneStep` that uses a different minibatching scheme (and plug this into a different `execution_plan`). Does that seem reasonable, or there an easier way to do this? Further, I'm not sure how I can get those `Tensor`s of shape `(N_windows, window_size, obs_size)` even if I get a set of same-length windows in the minibatch, because the `UsageTrackingDict` cannot be used to perform `split_by_episodes`.
